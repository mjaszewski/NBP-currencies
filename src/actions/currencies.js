export const GET_DATA_REQUESTED = "GET_DATA_REQUESTED";
export const GET_DATA_DONE = "GET_DATA_DONE";

export function getDataCurrenciesRequested(name) {
    return {
        type: GET_DATA_REQUESTED,
        payload: name,
    };
}

export function getDataCurrenciesDone(data, name) {
    return {
        type: GET_DATA_DONE,
        payload: {data, name},
    };
}