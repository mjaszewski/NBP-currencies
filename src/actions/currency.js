export const FETCH_DATA = 'FETCH_DATA';

export function fetchCurrency(data) {
    return {
        type: 'FETCH_DATA',
        payload: data
    };
}