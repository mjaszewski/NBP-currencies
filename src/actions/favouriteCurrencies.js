export const DATA_ADDED = 'DATA_ADDED';
export const DATA_REMOVE = 'DATA_REMOVE';
export const DATA_REMOVE_ALL = 'DATA_REMOVE_ALL';

export function addFavouriteCurrency(data) {
    return {
        type: DATA_ADDED,
        payload: data,
    };
}