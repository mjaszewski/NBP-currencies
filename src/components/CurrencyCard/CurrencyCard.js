import React, {Component} from 'react';
import PropTypes from "prop-types";PropTypes

import {Card, Button} from "semantic-ui-react";

class CurrencyCard extends Component {

    componentDidMount() {
    }

    handleDelete = (id) => {
        this.props.handleDelete(id)
    };

    render() {
        const { name, id, value} = this.props;
        return (
            <Card>
                <Card.Content>
                    <Card.Header>{name}</Card.Header>
                    <Card.Meta>{id}</Card.Meta>
                    <Card.Description>
                        <strong>{value}</strong>
                    </Card.Description>
                </Card.Content>
                <Card.Content extra>
                    <div className='ui one buttons'>
                        <Button basic negative onClick={this.handleDelete(id)}>
                            Remove
                        </Button>
                    </div>
                </Card.Content>
            </Card>
        );
    }
}

CurrencyCard.propTypes = {
    name: PropTypes.string,
    id: PropTypes.string,
    value: PropTypes.number,
    handleDelete: PropTypes.func

};

export default CurrencyCard