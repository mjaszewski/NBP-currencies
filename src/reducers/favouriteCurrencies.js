import {DATA_ADDED, DATA_REMOVE, DATA_REMOVE_ALL} from "../actions/favouriteCurrencies";

const initialState = [];

export default (state = initialState, action) => {
    switch (action.type) {
        case DATA_ADDED:
            return [...state, action.payload];
        case DATA_REMOVE:
            return state.filter(val => val.code !== action.payload);
        case DATA_REMOVE_ALL:
            return [];
        default:
            return state;
    }
}