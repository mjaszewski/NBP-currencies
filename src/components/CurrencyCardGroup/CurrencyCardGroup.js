import React, {Component} from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {toastr} from "react-redux-toastr";
import {Card, Button, Message} from "semantic-ui-react";
import CurrencyCard from "../CurrencyCard";
import {bindActionCreators} from "redux";
import {removeFavouriteCurrency, removeAllFavouriteCurrency} from "../../actions/favouriteCurrencies";


class CurrencyCardGroup extends Component {

    handleDelete = (code) => () => {
        this.props.removeFavouriteCurrency(code);
        toastr.success("Success", "Currency was successfully removed from favourites");
    };

    handleDeleteAll = () => {
        this.props.removeAllFavouriteCurrency();
        toastr.success("Success", "Currencies was successfully removed from favourites");
    };

    render() {
        const {favouriteCurrencies} = this.props;

        return (
            <div>
                <div className="delete-button-container">
                    <Button id="remove-all" negative disabled={favouriteCurrencies.length === 0} onClick={this.handleDeleteAll}>Remove
                        All</Button>
                </div>
                <Card.Group>
                    {favouriteCurrencies.length === 0 && (
                        <Message className="no-display-message">
                            <Message.Header>Nothing to display</Message.Header>
                        </Message>
                    )}
                    {favouriteCurrencies.map(currency => {
                            return (
                                <CurrencyCard key={currency.currency} name={currency.currency} code={currency.code}
                                              value={currency.mid} date={currency.date} handleDelete={this.handleDelete}/>
                            )
                        }
                    )}
                </Card.Group>
            </div>

        );
    }
}

function mapStateToProps(state) {
    return {
        favouriteCurrencies: state.favouriteCurrencies,
    }
}

const mapDispatchToProps = (dispatch) => {
    return bindActionCreators({
        removeFavouriteCurrency,
        removeAllFavouriteCurrency,
    }, dispatch);
};

CurrencyCardGroup.propTypes = {
    removeFavouriteCurrency: PropTypes.func,
    removeAllFavouriteCurrency: PropTypes.func,
    favouriteCurrencies: PropTypes.array
};

export default connect(mapStateToProps, mapDispatchToProps)(CurrencyCardGroup);
