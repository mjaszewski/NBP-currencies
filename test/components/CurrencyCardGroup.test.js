import {Provider} from "react-redux";
import configureStore from "redux-mock-store";
import CurrencyCardGroup from "../../src/components/CurrencyCardGroup/CurrencyCardGroup";
import mockedStore from "../mocks/store";
import CurrencyCard from "../../src/components/CurrencyCard";
import {removeAllFavouriteCurrency, removeFavouriteCurrency, DATA_REMOVE_ALL, DATA_REMOVE} from "../../src/actions/favouriteCurrencies"
const mockStore = configureStore();

let store;
let wrapper;
let group;

beforeEach(() => {
    store = mockStore(mockedStore);
    wrapper = mount(<Provider store={store}>
        <CurrencyCardGroup/>
    </Provider>);
    group = wrapper.find('CurrencyCardGroup').instance();
    store.clearActions();
});

describe("Render CurrencyCardGroup", () => {
    it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it("should render with given state from Redux store", () => {
        expect(wrapper.find(CurrencyCard)).toHaveLength(2);
    });
});

describe("CurrencyCardGroup interactions", () => {
    it("should call the handleDeleteAll function when \\'Remove all\\' button is clicked", () => {
        const mockedHandleClickRemoveAll = jest.fn();
        group.handleDeleteAll = mockedHandleClickRemoveAll;
        group.forceUpdate();
        wrapper.update();

        wrapper.find('button#remove-all').props().onClick();
        store.dispatch(removeAllFavouriteCurrency());
        const actions = store.getActions();
        const expectedPayload =  { type: DATA_REMOVE_ALL };

        expect(mockedHandleClickRemoveAll).toHaveBeenCalledTimes(1);
        expect(actions).toEqual([expectedPayload])
    });
});