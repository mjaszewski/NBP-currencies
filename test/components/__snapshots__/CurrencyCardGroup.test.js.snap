// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render CurrencyCardGroup renders correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(CurrencyCardGroup)>
    <CurrencyCardGroup
      favouriteCurrencies={
        Array [
          Object {
            "code": "THB",
            "currency": "bat (Tajlandia)",
            "date": "2020-02-21",
            "mid": 0.1252,
          },
          Object {
            "code": "USD",
            "currency": "dolar amerykański",
            "date": "2020-02-21",
            "mid": 3.9623,
          },
        ]
      }
      removeAllFavouriteCurrency={[Function]}
      removeFavouriteCurrency={[Function]}
    >
      <div>
        <div
          className="delete-button-container"
        >
          <Button
            as="button"
            disabled={false}
            id="remove-all"
            negative={true}
            onClick={[Function]}
          >
            <Ref
              innerRef={
                Object {
                  "current": <button
                    class="ui negative button"
                    id="remove-all"
                  >
                    Remove All
                  </button>,
                }
              }
            >
              <RefFindNode
                innerRef={
                  Object {
                    "current": <button
                      class="ui negative button"
                      id="remove-all"
                    >
                      Remove All
                    </button>,
                  }
                }
              >
                <button
                  className="ui negative button"
                  id="remove-all"
                  onClick={[Function]}
                >
                  Remove All
                </button>
              </RefFindNode>
            </Ref>
          </Button>
        </div>
        <CardGroup>
          <div
            className="ui cards"
          >
            <CurrencyCard
              code="THB"
              date="2020-02-21"
              handleDelete={[Function]}
              key="bat (Tajlandia)"
              name="bat (Tajlandia)"
              value={0.1252}
            >
              <Card>
                <div
                  className="ui card"
                  onClick={[Function]}
                >
                  <CardContent>
                    <div
                      className="content"
                    >
                      <CardHeader>
                        <div
                          className="header"
                        >
                          bat (Tajlandia)
                           - 
                          2020-02-21
                        </div>
                      </CardHeader>
                      <CardMeta>
                        <div
                          className="meta"
                        >
                          THB
                        </div>
                      </CardMeta>
                      <CardDescription>
                        <div
                          className="description"
                        >
                          <strong>
                            0.1252
                             zł
                          </strong>
                        </div>
                      </CardDescription>
                    </div>
                  </CardContent>
                  <CardContent
                    extra={true}
                  >
                    <div
                      className="extra content"
                    >
                      <div
                        className="ui one buttons"
                      >
                        <Button
                          as="button"
                          basic={true}
                          className="remove-favourite"
                          negative={true}
                          onClick={[Function]}
                        >
                          <Ref
                            innerRef={
                              Object {
                                "current": <button
                                  class="ui basic negative button remove-favourite"
                                >
                                  Remove
                                </button>,
                              }
                            }
                          >
                            <RefFindNode
                              innerRef={
                                Object {
                                  "current": <button
                                    class="ui basic negative button remove-favourite"
                                  >
                                    Remove
                                  </button>,
                                }
                              }
                            >
                              <button
                                className="ui basic negative button remove-favourite"
                                onClick={[Function]}
                              >
                                Remove
                              </button>
                            </RefFindNode>
                          </Ref>
                        </Button>
                      </div>
                    </div>
                  </CardContent>
                </div>
              </Card>
            </CurrencyCard>
            <CurrencyCard
              code="USD"
              date="2020-02-21"
              handleDelete={[Function]}
              key="dolar amerykański"
              name="dolar amerykański"
              value={3.9623}
            >
              <Card>
                <div
                  className="ui card"
                  onClick={[Function]}
                >
                  <CardContent>
                    <div
                      className="content"
                    >
                      <CardHeader>
                        <div
                          className="header"
                        >
                          dolar amerykański
                           - 
                          2020-02-21
                        </div>
                      </CardHeader>
                      <CardMeta>
                        <div
                          className="meta"
                        >
                          USD
                        </div>
                      </CardMeta>
                      <CardDescription>
                        <div
                          className="description"
                        >
                          <strong>
                            3.9623
                             zł
                          </strong>
                        </div>
                      </CardDescription>
                    </div>
                  </CardContent>
                  <CardContent
                    extra={true}
                  >
                    <div
                      className="extra content"
                    >
                      <div
                        className="ui one buttons"
                      >
                        <Button
                          as="button"
                          basic={true}
                          className="remove-favourite"
                          negative={true}
                          onClick={[Function]}
                        >
                          <Ref
                            innerRef={
                              Object {
                                "current": <button
                                  class="ui basic negative button remove-favourite"
                                >
                                  Remove
                                </button>,
                              }
                            }
                          >
                            <RefFindNode
                              innerRef={
                                Object {
                                  "current": <button
                                    class="ui basic negative button remove-favourite"
                                  >
                                    Remove
                                  </button>,
                                }
                              }
                            >
                              <button
                                className="ui basic negative button remove-favourite"
                                onClick={[Function]}
                              >
                                Remove
                              </button>
                            </RefFindNode>
                          </Ref>
                        </Button>
                      </div>
                    </div>
                  </CardContent>
                </div>
              </Card>
            </CurrencyCard>
          </div>
        </CardGroup>
      </div>
    </CurrencyCardGroup>
  </Connect(CurrencyCardGroup)>
</Provider>
`;
